aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
   return std::tuple<Tensor&, Tensor&>(vals, vecs);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ linalg_eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+// This function returns complex-valued eigenvectors that is obtained from LAPACK GEEV's real-valued output
+// This function is also used for the MAGMA path because intermediate MAGMA's results live on CPU
+template <typename scalar_t>
+static void linalg_eig_make_complex_eigenvectors_impl(Tensor& result, const Tensor& complex_values, const Tensor& real_vectors) {
+  // From GEEV documentation:
+  // Complex conjugate pairs of eigenvalues appear consecutively with the eigenvalue having the positive imaginary part first
+  // If the j-th eigenvalue is real, then v(j) = VR(:,j), the j-th column of VR.
+  // If the j-th and (j+1)-st eigenvalues form a complex conjugate pair, then v(j) = VR(:,j) + i*VR(:,j+1) and v(j+1) = VR(:,j) - i*VR(:,j+1).
+
+  auto batch_size = batchCount(real_vectors);
+  auto n = real_vectors.size(-1);
+  auto matrix_stride = matrixStride(real_vectors);
+
+  auto result_data = result.data_ptr<c10::complex<scalar_t>>();
+  auto real_vectors_data = real_vectors.data_ptr<scalar_t>();
+  auto values_data = complex_values.data_ptr<c10::complex<scalar_t>>();
+
+  for (auto b = decltype(batch_size){0}; b < batch_size; b++) {
+    scalar_t* vecs = &real_vectors_data[b * matrix_stride];
+    c10::complex<scalar_t>* res = &result_data[b * matrix_stride];
+    c10::complex<scalar_t>* vals = &values_data[b * n];
+    for (auto j = decltype(n){0}; j < n; j++) {
+      if (vals[j].imag() == 0.0) {  // eigenvalue is real, then v(j) = VR(:,j)
+        for (auto i = decltype(n){0}; i < n; i++) {
+          res[j * n + i] = c10::complex<scalar_t>(vecs[j * n + i], 0);
+        }
+      } else {
+        for (auto i = decltype(n){0}; i < n; i++) {
+          res[j * n + i] = c10::complex<scalar_t>(vecs[j * n + i],  vecs[(j+1) * n + i]);      // v(j)   = VR(:,j) + i*VR(:,j+1)
+          res[(j+1) * n + i] = c10::complex<scalar_t>(vecs[j * n + i], -vecs[(j+1) * n + i]);  // v(j+1) = VR(:,j) - i*VR(:,j+1)
+        }
+        j++;
+      }
+    }
+  }
+}
```

Comment:
You can make this code device-independent with eigenvalue synchronization if on CUDA by doing something along these lines:
```
res_3d = res.view({-1, res.size(-2), res.size(-1)});
values_2d = values.view({-1, values.size(-1)});
real_vectors_3d = real_vectors.view({-1, real_vectors.size(-2), real_vectors.size(-1)});
for (b = ....) {
   res = res_3d.select(0, b);
   values = values_2d.select(0, b);
   real_vectors = real_vectors_3d.select(0, b):
   for (j = ....) {
       res.select(-1, j) = real_vectors.select(-1, j) + i * real_vectors.select(-1, j + 1);
   }
}
```

Reply:
Yes, I have the following prototype working
```
  result.copy_(real_vectors);
  auto conjugate_pair_indices = (at::imag(complex_values) > 0).nonzero().split(1, 1);
  auto shifted_indices = conjugate_pair_indices;
  shifted_indices.back() = shifted_indices.back() + 1;
  auto result_imag = at::imag(result);
  auto indices = at::native::toListOfOptionalTensors(conjugate_pair_indices);
  auto shifted_indices_ = at::native::toListOfOptionalTensors(shifted_indices);
  result_imag.transpose(-2, -1).index_put_(std::move(indices), real_vectors.transpose(-2, -1).index(std::move(shifted_indices_)));
  result.transpose(-2, -1).index_put_(std::move(shifted_indices_), result.transpose(-2, -1).index(indices).conj());
  return result;
  ```
  It didn't give any speedup when I checked, maybe some things could be made faster.
  But here we always have contiguous inputs that are always on the CPU and no computations just moving the memory around, so the compiler should make it efficient.

Reply:
Something like this could probably work:
```
auto idx_pure_real = (at::imag(values_2d) == 0.0).nonzero().squeeze(-1) // returns 2d tensor, but we need 1d
res_3d.index_put_({idx_pure_real}, at::index_select(real_vectors_2d, 0, idx_pure_real));
```

Reply:
I also suspect that you can do this without even the per-matrix loop over `for(b = ...)`.